### ISSUE TITLE
Randomly slow youtube download speed 

### ISSUE THREAD

**Issue by a26364177** [fent/node-ytdl-core#964]
Hello,

Important: The problem is random, maybe 1 chance on 8 to produce it. You have to download several videos in a row (about 10) to notice it.

Since a few weeks, randomly a youtube video can be slowed down to 48 Kio/s, so it takes 5-10 minutes to download a short video of 5 minutes instead of 4, 5 seconds, often the download does not succeed and stops after a few minutes.

This happens on several servers, several internet providers as well as with my private connection.

base on this: https://github.com/ytdl-org/youtube-dl/issues/29326

**Comment by redbrain** [comment]
I've experienced this with other ytdl libraries as well, such as the one linked above.
If they find a way to fix it, I propose we implement the same. Until then, I don't think there's much to do.

**Comment by gatecrasher777** [comment]
Is it unique to specific videos?  If you download the same video again later does it still download slowly or at the normal speed? 

If it is repeatable, you could try reducing dlChunkSize (defaults to 10MB) to see whether more frequent calls for smaller chunks improves the download speed. 

Lowering the dlChunkSize will reduce the normal download speed by a fraction, but if doing so can avoid the randomly slow downloads, it might be worth it.

**Comment by gatecrasher777** [comment]
Been looking at this. What I have found is that when there is a slow download, and I cancel downloading and restart some seconds later it then downloads at full steam.

So what is needed is some kind of download speed threshold below which the download is cancelled and after a short pause restarted.  But it also needs a counter, so that the number of restarts is limited to prevent an infinite restart loop. 

In the thread [https://github.com/ytdl-org/youtube-dl/issues/29326](url) there is the suggestion of manipulation of the n parameter, but if you look at the Dea function it is extremely obfuscated, much more so than the decipher methods.  I would think you would have to be execute the original code in base.js directly for this to be a solution.

**Comment by hwgilbert16** [comment]
I’ve encountered the same issue as well in my development of [youtube-archiver](https://github.com/hwgilbert16/youtube-archiver). I initially thought that the issue was isolated to a codec, and I eliminated the one problem child, but then I saw it randomly happened to any of them.

My best guess is it may be something on YouTube’s end, perhaps it finds something fishy with the request to googlevideo and ratelimits it to prevent a bot attack. The best mitigation for now it seems is to monitor the readable stream. If the completion rate is too slow, cancel the download and restart it.

**Comment by LuanRT** [comment]
This was giving me a lot of headaches a while ago. I found a way to mitigate it doe, not the most elegant way but it works fine for me. However, if you're on a slow internet connection you might end up in an infinite loop, so I suggest you to play around with the code a bit. 

```js
const downloadVideo = (link) => {
  const output_video = './temp/video.mp4';
  return new Promise(function (resolve, reject) {
    const start = () => {
      const stream = ytdl(link, {
        filter: (format) => format.container === "mp4",
        requestOptions: {
          headers: {
            cookie: YT_COOKIE,
          },
        },
      });

      stream.pipe(fs.createWriteStream(output_video));

      stream.once("response", () => {
        starttime = Date.now();
      });

      stream.on("progress", (chunkLength, downloaded, total) => {
        const percent = downloaded / total;
        const downloaded_minutes = (Date.now() - starttime) / 1000 / 60;
        const estimated_download_time = downloaded_minutes / percent - downloaded_minutes;

        // if the estimated download time is more than 1.5 minutes then we cancel and restart the download, this value works fine for me but you may need to change it based on your server/internet speed.
        if (estimated_download_time.toFixed(2) >= 1.5) {
          console.warn("Seems like YouTube is limiting our download speed, restarting the download to mitigate the problem..");
          stream.destroy();
          start();          
        }
      });

      stream.on("end", () => {
        resolve(output_video);
      });

      stream.on("error", (error) => {
        reject(error);
      });
    };
    start();
  });
};
```

**Comment by gatecrasher777** [comment]
To overcome this throttling problem, ytdl-core may have to run the n transform function in the base.js player code. I do this in my [innertube library](https://github.com/gatecrasher777/ytcog) and do not experience any throttled downloads. The n transform is a lot more complex than the signature decipher function. I've seen other libraries try to emulate the n transform but it differs substantially from yt player to yt player, so this approach is very cumbersome and not very reliable.  Another successful approach I have seen is to emulate an android phone, where this throttling is apparently not being applied (perhaps a reward for using Google's OS).

**Comment by VVD** [comment]
Both yt-dlp and youtube-dl are constantly slow last several days.

**Comment by oceangem** [comment]
Running into the same problem :( consistent 50-60 kbps download speeds, even after setting a higher rate via the -r flag.

**Comment by LuanRT** [comment]
> This problem used to be random, but now seems to be on every video I try without solving the n-transform challenge.
> 
> I have made a PR. Hope it can be accepted. The alternative to running the Google code directly is to attempt to emulate the n-parameter challenge code. But it is hugely more complex than the decipher code and way above my ability.
> 
> Looking around at other YT projects, those emulation attempts have generally worked for a few days and then failed with newer player versions.
> 
> Projects that are working around the problem are using the ANDROID fix, but then they are having to jump through some hoops to overcome the limitations of the ANDROID fix. And it seems almost inevitable to me that Google will close this bypass soon.

Yup, same,
Was testing ytdl here too and now every video is being throttled. To be honest executing the N transformation algorithm directly is the only way to fix it, as you said, emulating it won't work because it changes every time a new player is released.

**Comment by VVD** [comment]
> > Yup I have the same issue, the download speed never exceeds 82kB/s which is very problematic
> > I know that 2 weeks ago, I already have noticed that it was kinda slow, but when I stopped, waited 5-10 sec and then redid the downloading, it was going fast again. However, this technique is not working anymore..
> 
> [#964 (comment)](https://github.com/fent/node-ytdl-core/issues/964#issuecomment-945333981) You need to use yt-dlp now.

yt-dlp affected too. :-(

**Comment by hwgilbert16** [comment]
This is likely caused by an update to YouTube's API.

**Comment by gatecrasher777** [comment]
> IDM downloads the same video much faster, but always with a constant rate (~700 KB/s), sometimes faster. 

That's because IDM has multithreaded downloads, so downloading 10 segments simultaneously = 10x70 KB/s =700 KB/s. 

The innertube library [ytcog](https://github.com/gatecrasher777/ytcog) and CLI [ytcog-dl](https://github.com/gatecrasher777/ytcog-dl) are unthrottled using the same n-parameter solution as proposed in PR #1022.  Might help some people while they wait for a fix.

**Comment by stefankiebacher** [comment]
Same in Karlsruhe (Germany)  between 50-80 KiB/s though 100MiB stable connection.
`[download]  75.5% of 2.28MiB at 52.22KiB/s ETA 00:10`

**Comment by itseyup** [comment]
Same in my United States server. between 50-80 KiB/s though 2000MiB stable connection.

**Comment by hangat** [comment]
In my case, speed > 1 MB/s for Video Part (.webm) and speed drop to <= 64KiB/s for Audio Part (.m4a)

**Comment by blackjyn** [comment]
same here , it stucks at 40-50KiB/s with 1000+MiB super stable connection.
I think it is a bug

**Comment by Sansekai** [comment]
now version 4.10.0 too slow

**Comment by Cherubrine** [comment]
Same for me since yesterday evening (Austria (Main IP) / Germany (ISP Proxies) / Netherlands (ISP Proxies))

**Comment by gatecrasher777** [comment]
This is a new but small obstacle from YouTube in their base player files. Will put up a PR later.

**Comment by TimeForANinja** [comment]
yep, fixed with #1055 
gonna freeze this now - so pls open a new issue when experiencing this issue again
